File: 52-n-queens-ii/n-queens-ii.ts
function totalNQueens(n: number): number {
    let count = 0;

    // Sets to track occupied cols and diagonals
    const cols = new Set<number>();
    const diag1 = new Set<number>(); // row - col
    const diag2 = new Set<number>(); // row + col

    function backtrack(row: number) {
        if (row === n) {
            count++;
            return;
        }

        for (let col = 0; col < n; col++) {
            if (cols.has(col) || diag1.has(row - col) || diag2.has(row + col)) {
                continue;
            }

            // Place queen
            cols.add(col);
            diag1.add(row - col);
            diag2.add(row + col);

            backtrack(row + 1);

            // Backtrack
            cols.delete(col);
            diag1.delete(row - col);
            diag2.delete(row + col);
        }
    }

    backtrack(0);
    return count;
}

File: 52-n-queens-ii/n-queens-ii.cpp
class Solution {
public:
    int totalNQueens(int n) {
        // sets to track attacked columns and diagonals
        unordered_set<int> cols, diag1, diag2;
        int count = 0;

        function<void(int)> backtrack = [&](int row) {
            if (row == n) {
                count++;
                return;
            }

            for (int col = 0; col < n; col++) {
                if (cols.count(col) || diag1.count(row - col) ||
                    diag2.count(row + col))
                    continue;

                // place queen
                cols.insert(col);
                diag1.insert(row - col);
                diag2.insert(row + col);

                backtrack(row + 1);

                // remove queen (backtrack)
                cols.erase(col);
                diag1.erase(row - col);
                diag2.erase(row + col);
            }
        };

        backtrack(0);
        return count;
    }
};
